giraph-core/src/main/java/org/apache/giraph/ooc/policy/MemoryEstimatorOracle.java

```diff
@@ -0,0 +1,851 @@
+            for (int col = i + 1; col < 5; ++col) {
+              if (isLinearDependence(dataSamples, i, col)) {
+                validIndex = false;
+                break;
+              }
+            }
+          }
+        }
+
+        if (validIndex) {
+          validColumnIndices.add(i);
+        }
+      }
+
+      // If we filtered out columns in the previous step, we are going to run
+      // the regression without those columns.
+
+      // Create the coefficient table
+      boolean setIsValid = false;
+      lock.lock();
+      try {
+        if (validColumnIndices.size() >= 1 &&
+          dataSamples.size() >= validColumnIndices.size() + 1) {
+
+          double[][] xValues = new double[dataSamples.size()][];
+          double[] yValues = new double[memorySamples.size()];
+          fillXMatrix(dataSamples, validColumnIndices, xValues);
+          copyVectorToArray(memorySamples, yValues);
+          mlr.newSampleData(yValues, xValues);
+          calculateRegression(coefficient, validColumnIndices, mlr);
+
+          // After the computation of the regression, some coefficients may have
+          // values outside the valid value range. In this case, we set the
+          // coefficient to the minimum or maximum value allowed, and re-run the
+          // regression.
+          boolean changed;
+          extreme[3] = -1;
+          extreme[4] = -1;
+          do {
+            changed = refineCoefficient(4, 1, 2, xValues, yValues);
+            changed |= refineCoefficient(3, 0, 2, xValues, yValues);
+          } while (changed);
+          if (extreme[3] != -1) {
+            coefficient[3] = extreme[3];
+          }
+          if (extreme[4] != -1) {
+            coefficient[4] = extreme[4];
+          }
+          setIsValid = true;
+          return; // the finally-block will execute before return
+        }
+        // CHECKSTYLE: stop IllegalCatch
+      } catch (Exception e) {
+        // CHECKSTYLE: resume IllegalCatch
+        LOG.warn("addRecord: exception occurred!", e);
+      } finally {
+        // This inner try-finally block is necessary to ensure that the
+        // lock is always released.
+        try {
+          isValid = setIsValid;
+          printStats();
+        } finally {
+          lock.unlock();
+        }
+      }
+    }
+
+    /**
+     * Certain coefficients need to be within a specific range. For instance,
+     * If the coefficient is not in this range, we set it to the closest bound
+     * and re-run the linear regression.
+     *
+     * @param coefIndex Coefficient index
+     * @param lowerBound Lower bound
+     * @param upperBound Upper bound
+     * @param xValues double[][] matrix with data samples
+     * @param yValues double[] matrix with y samples
+     * @return True if coefficients were out-of-range
+     * @throws Exception
+     */
+    private boolean refineCoefficient(int coefIndex, double lowerBound,
```

Comment:
Actually, is there a way to do regularization in this Java library? That would be a nice way to make sure you're coefficients stay within a reasonable range